---
title: "An analysis of cutaneous leishmaniasis biopsies: `r Sys.getenv('VERSION')`"
author: "Najib El-Sayed"
date: "`r Sys.Date()`"
output:
  html_document:
    code_download: true
    code_folding: show
    fig_caption: true
    fig_height: 7
    fig_width: 7
    highlight: zenburn
    keep_md: false
    mode: selfcontained
    number_sections: true
    self_contained: true
    theme: readable
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: false
---

<style type="text/css">
body, td {
  font-size: 16px;
}
code.r{
  font-size: 16px;
}
pre {
  font-size: 16px
}
body .main-container {
  max-width: 1600px;
}
</style>


```{r options, include=FALSE}
library(hpgltools)
library(glue)
knitr::opts_knit$set(
  progress = TRUE, verbose = TRUE, width = 90, echo = TRUE)
knitr::opts_chunk$set(
  error = TRUE, fig.width = 8, fig.height = 8, fig.retina = 2,
  fig.pos = "t", fig.align = "center", dpi = if (knitr::is_latex_output()) 72 else 300,
  out.width = "100%", dev = "png",
  dev.args = list("png" = list(type = "cairo-png")))
old_options <- options(
  digits = 4, stringsAsFactors = FALSE, knitr.duplicate.label = "allow")
ggplot2::theme_set(ggplot2::theme_bw(base_size = 12))
ver <- Sys.getenv("VERSION")
rundate <- format(Sys.Date(), format = "%Y%m%d")

rmd_file <- "01index.Rmd"
savefile <- gsub(pattern = "\\.Rmd", replace = "\\.rda\\.xz", x = rmd_file)
```

# Consolidation of a Molecular Signature of Healing in Cutaneous Leishmaniasis is Achieved During the First Ten Days of Treatment.

# Introduction

This document performs a series of diagnostics and differential
expression analyses which eventually helped inform the text of Lina
Fernanda Giraldo Parra's and Maria Adelaida Gomez's paper.  In it,
they sought to use biopsies over time from people who were observed to
cure/fail treatment of cutaneous leishmaniasis.  Some of these people
were treated with the antimonial 'glucantime', others with
miltefosine.  Biopsies were taken at the beginning, middle, and end of
treatment.  RNA was taken from the biopsies, made into libraries, and
sequenced.  The samples were mapped/counted against ensembl hg38
release 100 and the TritrypDB Leishmania panamensis MHOM/COL strain's
genome with hisat2 and salmon.  There were very few reads observed
from the parasite.  As a result (at the time of this writing: 202310),
those count tables are not included in the sample sheet (they are now,
along with the salmon quantitations against hg38_100), and are not
considered in this worksheet.

# Changelog

* 20231025: All the PCA plots were of non-filtered data.  I removed
  two samples due to coverage, so the plots should not
  include them.  While I am at it, rename a few variables to make
  everything more consistent.  Also explicitly writing the
  figure-pieces to the 'images/' directory, previously I just created
  them on request.
* 202309-202310: Copied the worksheet from my directory into the
  container build tree, reorganized it, and added some text to
  hopefully address queries/concerns from reviewers as well as clarify
  some tasks performed.
* 2022-2023: Copied Najib's worksheet to my directory, cleaned it up a
  bit, and finished what I think are the likely analyses.

I collected the available metadata into the sheet 'all_samples.xlsx'.

# Sample sheet

```{r}
samplesheet <- "sample_sheets/all_samples.xlsx"
```

## Extracting extra sample annotations

The function 'gather_preprocessing_metadata()' scans the working
directory in which the various preprocessing tasks were performed,
trimming, fastqc/fastp, mapping/quantitation, variant searching, etc.
It is aware of my default output filenames for stdout/stderr/etc and
extracts from them various metrics potentially of interest.  I think
the sample sheet I copied into this container is the result of the
following block, so I am not going to evaluate it here.

```{r gather_preprocessing, eval=FALSE}
rna_spec <- make_rnaseq_spec()
modified <- gather_preprocessing_metadata(samplesheet,
                                          specification = rna_spec,
                                          species = "hg38_100")
```

# Annotation

We take the annotation data from ensembl's biomart instance.  The genome which
was used to map the data was hg38 revision 100.  My default when using biomart is
to load the data from 1 year before the current date.

```{r}
hs_annot <- load_biomart_annotations(year = "2019")
hs_annot

hs_annot <- hs_annot[["annotation"]]
hs_annot[["transcript"]] <- paste0(rownames(hs_annot), ".", hs_annot[["version"]])
rownames(hs_annot) <- make.names(hs_annot[["ensembl_gene_id"]], unique = TRUE)
tx_gene_map <- hs_annot[, c("transcript", "ensembl_gene_id")]
```

## Gene Ontology data

Downloading the Ontology annotation information has a significant
chance of failing because it sometimes takes longer than curl's
timeout of 300 seconds.  As a result, I may decide to include the
saved rda of ontology results in the container, or maybe improve my
ontology downloader so that it is no longer susceptible to timeouts...

Actually, I do not think I am using
clusterProfiler/gostats/goseq/topgo in this document, perhaps
therefore I should just exclude this block?  Yeah, it seems like there
is a 50/50 chance that this will fail when run within the container,
and I am currently not using this set of gene:GO mappings in this document.

```{r, eval=FALSE}
hs_go <- load_biomart_go(overwrite = TRUE)
hs_go
hs_go <- hs_go[["go"]]
hs_length <- hs_annot[, c("ensembl_gene_id", "cds_length")]
colnames(hs_length) <- c("ID", "length")
```

# Create expressionsets

The primary datastructure used throughout this document is the
expressionSet or summarized Experiment.  This document uses the
former; but both are created by reading the sample sheet, extracting
the filenames of the count data from it, and combining the metadata,
annotation data, and counts.  My implementation of this process also
includes a few extra pieces of information.  For example: the colors
of various potential conditions in the data.

## Color choices

Before we create the datastructures, I want to have the various color
schemes in place.  Many of these color choices are taken directly from
another project which also looks at host responses.

```{r}
color_choices <- list(
  "cf_visit" = list(
    "cure_v1" = "#D95F0E",
    "failure_v1" = "#FEC44F",
    "cure_v2" = "#DD1C77",
    "failure_v2" = "#C994C7",
    "cure_v3" = "#3182BD",
    "failure_v3" = "#9ECAE1"),
  "drug_visit" = list(
    "antimoniate_v1" = "#badeef",
    "miltefosine_v1" = "#e4a6c1",
    "antimoniate_v2" = "#7bc6ea",
    "miltefosine_v2" = "#d26895",
    "antimoniate_v3" = "#1d91ca",
    "miltefosine_v3" = "#d52e75"),
  "species_visit" = list(
    "lvpanamensis_v1" = "#ffe798",
    "lvbraziliensis_v1" = "#b5b5b5",
    "lvpanamensis_v2" = "#FFC300",
    "lvbraziliensis_v2" = "#888888",
    "lvpanamensis_v3" = "#b58b00",
    "lvbraziliensis_v3" = "#525252"),
  "cf" = list(
    "cure" = "#998EC3",
    "failure" = "#F1A340"),
  "visit" = list(
    "v1" = "#33EE33",
    "v2" = "#11AA11",
    "v3" = "#134413"),
  "visitbi" = list(
    "v1" = "#BB0000",
    "vother" = "#0000BB"),
  "species" = list(
    "lvpanamensis" = "#FFC300",
    "lvbraziliensis" = "#525252"),
  "donor" = list(
    "d2008" = "#B78415",
    "d1029" = "#93752C",
    "d1036" = "#7E6EA2",
    "d1037" = "#B3499C",
    "d1031" = "#BD6332",
    "d2002" = "#7D8F31",
    "d2009" = "#8E7037",
    "d2010" = "#666666",
    "d1019" = "#1B9E77",
    "d2004" = "#E0A604",
    "d2001" =  "#CF3F76",
    "d2003" = "#A0A811"),
  "drug" = list(
    "antimoniate" = "#3182AA",
    "miltefosine" = "#C994AA"))
```

The set of color choices demonstrates the complexity of the
experimental design.  We are looking at combinations of time, outcome,
species, and drug.  Inherent in all of these is an overarching donor
effect.

## Initial dataset: Combine clinical outcome with visit

The initial datastructure will use a factor combining the clinical
outcome and visit as the experimental condition of interest.  We will
follow this up by splitting off the various factors of interest.

An aside: when testing the following block within the container, it
did not print out the filenames as it was reading them, usually it
does.  I assume this is not a problem since the data looks fine, but
this may warrant further exploration.

The call to sanitize_expt_pData() tries to ensure that Excel does not
mess things up with random spaces/capitalization in the metadata.

```{r}
wellcome_outtime <- create_expt(metadata = samplesheet, gene_info = hs_annot,
                                file_column = "hg38100hisatfile") %>%
  set_expt_batches(fact = "drug") %>%
  sanitize_expt_pData(spaces = TRUE)
wellcome_outtime
```

The previous block creates the primary datastructure, everything from
now on will either:

* Add categories to it (the next block)
* Set various categories to the primary factor of interest
  (set_expt_conditions()/set_expt_batches())
* Subset the data to highlight individual questions or remove the two
  (I think two, double check this) low-coverage samples.
* Normalize/analyze the above.

The following block therefore creates a set of factors which are the
concatenation of time+x where x is one of: outcome, drug, species.

```{r}
outcome_time <- paste0(pData(wellcome_outtime)[["clinicaloutcome"]], "_v",
                       pData(wellcome_outtime)[["visitnumber"]])
wellcome_outtime <- set_expt_conditions(wellcome_outtime, fact = outcome_time,
                                        colors = color_choices[["cf_visit"]])

pData(wellcome_outtime)[["visitnumber"]] <- paste0("v", pData(wellcome_outtime)[["visitnumber"]])
pData(wellcome_outtime)[["outcome_visit"]] <- outcome_time
pData(wellcome_outtime)[["drug_visit"]] <- paste0(pData(wellcome_outtime)[["drug"]], "_",
                                                  pData(wellcome_outtime)[["visitnumber"]])
pData(wellcome_outtime)[["species_visit"]] <- paste0(pData(wellcome_outtime)[["infectingspecies"]], "_",
                                                     pData(wellcome_outtime)[["visitnumber"]])
```

## Check samples

First, let us get a quick view of the samples in their native state.
There are a few samples which are candidates for removal due to lower
coverage.

Random aside: I have some nifty code which tries to autodetect when to
color the text in the colored bars white or black.  The purple/pink
color tricks that code into thinking it is dark enough to be colored
white.

* FIXME plot_libsize() text color heuristic.

```{r}
plot_legend(wellcome_outtime)
plot_libsize(wellcome_outtime)
plot_nonzero(wellcome_outtime, plot_labels = "repel")

wellcome_filtered <- subset_expt(wellcome_outtime, nonzero = 15000)
```

## Write all the sample data

The following writes out the data in two files, one before and after
filtering for samples with less than 15,000 observed genes.

```{r}
written <- write_expt(wellcome_outtime,
                      excel = glue("excel/CL_biopsies_all_data-v{ver}.xlsx"))
written <- write_expt(wellcome_filtered,
                      excel = glue("excel/CL_biopsies_filtered_data-v{ver}.xlsx"))
```

Looking at the nonzero plot, it seems that 15k genes is a reasonable
cutoff, which would remove 2 samples.  So, for the moment I will split
the data up into two sets, pre/post removal.

## Upload CPM values

Lina asked for a CSV copy of the data as CPM.

```{r}
wellcome_cpm <- normalize_expt(wellcome_filtered, filter = TRUE, convert = "cpm")
wellcome_cpm_mtrx <- exprs(wellcome_cpm)
wellcome_cpm_write <- write.csv(x = wellcome_cpm_mtrx, file = "excel/CL_biopsies_filtered_cpm.csv")

wellcome_scpm <- normalize_expt(wellcome_filtered, filter = TRUE, convert = "cpm", batch = "svaseq")
wellcome_scpm_mtrx <- exprs(wellcome_scpm)
wellcome_scpm_write <- write.csv(x = wellcome_cpm_mtrx, file = "excel/CL_biopsies_filtered_cpm_sva.csv")
```

## Check samples

We have in place some initial datastructures; let us mix and match
them to get a sense of the data.  There are a few factors in the
experiment that are likely of primary interest: the donor, time, drug
used, parasite species/strain, and clinical outcome.  There are too
many factors for the number of samples to do a full rank model, so I
am going to make copies of the expressionset and model each factor
separately with the help of sva.

## Only donor

Start with the donor.  There are 12 people, most of whom provided
three samples. Two of them (d2002 and d2009) had samples which got
excluded and so have only 2.  I am separating the un/filtered datasets
so that we may play with both if we want.

```{r}
wellcome_donor <- set_expt_conditions(wellcome_outtime, fact = "donor",
                                      colors = color_choices)

wellcome_donor_filt <- set_expt_conditions(wellcome_filtered, fact = "donor",
                                           colors = color_choices)
```

## Only visit

In this case, the condition will only be visit number.  We have 12
samples for each visit, except v1/v3 which are missing 1 each.

```{r}
wellcome_time <- set_expt_conditions(wellcome_outtime, fact = "visitnumber",
                                     colors = color_choices[["visit"]])

wellcome_time_filt <- set_expt_conditions(wellcome_filtered, fact = "visitnumber",
                                         colors = color_choices[["visit"]])
```

## Only Cure/Fail

Conversely, we can split by clinical outcome.  The unfiltered dataset
has 15 cures and 21 failures; both filtered samples were failure.

```{r}
wellcome_outcome <- set_expt_conditions(wellcome_outtime, fact = "clinicaloutcome",
                                        colors = color_choices[["cf"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_outcome_filt <- set_expt_conditions(wellcome_filtered, fact = "clinicaloutcome",
                                             colors = color_choices[["cf"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Only drug

With respect to drug treatment, we started with 18 of each.  The lost
samples were both miltefosine.

```{r}
wellcome_drug <- set_expt_conditions(wellcome_outtime, fact = "drug",
                                     colors = color_choices[["drug"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_drug_filt <- set_expt_conditions(wellcome_filtered, fact = "drug",
                                          colors = color_choices[["drug"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Only parasite

We started with 12 braziliensis and 24 panamensis; one of each was lost.

```{r}
wellcome_parasite <- set_expt_conditions(wellcome_outtime, fact = "infectingspecies",
                                         colors = color_choices[["species"]]) %>%
  set_expt_batches(fact = "visitnumber")

wellcome_parasite_filt <- set_expt_conditions(wellcome_filtered, fact = "infectingspecies",
                                              colors = color_choices[["species"]]) %>%
  set_expt_batches(fact = "visitnumber")
```

## Visualize the metadata

Everything I wrote above is visible in the following sankey plot.  It
is particularly noteworthy that we have no cure braziliensis for
miltefosine treatment and only 1 for each of the antominial
timepoints.

```{r}
drug_visit_species_cf <- plot_meta_sankey(
  wellcome_filtered, factors = c("drug", "visitnumber", "clinicaloutcome", "infectingspecies"),
  color_choices = color_choices)
drug_visit_species_cf
```

# Visualize samples

Given the large number of variables in this data, lets start by
getting a feeling for the amount of variance in each.  Given that we
don't have a full rank with respect to clinical outcome, I assume that
trying variance partition with the 4 primary factors will fail, but
let us see...

```{r}
varpart_donor <- simple_varpart(
  wellcome_filtered,
  factors = c("donor", "visitnumber"))
varpart_donor

varpart_dvic <- simple_varpart(
  wellcome_filtered,
  factors = c("drug", "visitnumber", "infectingspecies", "clinicaloutcome"))
varpart_dvic
```

To my eyes, it appears that donor is dominant factor,
followed by: visit, species, drug, and outcome.  I think this
observation may have an effect on the current state of the
manuscript - which if I recall properly states that donor is not a
significant factor in the data.

## Examine top-n genes with respect to variance of some factors

Let us ask if there are categories associated with the genes
associated with each of these categories and see if they 'make sense'.

```{r}
top_300_donor_genes_idx <- order(varpart_donor[["fitted_df"]][["donor"]])
top_300_donor_genes <- tail(varpart_donor[["fitted_df"]][top_300_donor_genes_idx, ], n = 300)
summary(top_300_donor_genes)
donor_genes_gp <- simple_gprofiler(rownames(top_300_donor_genes))
donor_genes_gp
donor_genes_gp$pvalue_plots$REAC
```

So, the top-300 genes with respect to putative donor effect, when
passed to gprofiler's over representation analyses, look like they
have lots of variance related to categories that are explicitly
important to the general questions we are asking.  I think that counts
as: Yay!

Now let us look at the other factors in the data and see if anything
noteworthy pops out.  Given that the dvic datastructure comprises the
other factors of interest, we will need to reorder the results with
respect to each factor separately.

### Visit

Visit also is dominated by variance which is explicitly what one would expect:
wound healing.

```{r}
top_300_visit_genes_idx <- order(varpart_dvic[["fitted_df"]][["visitnumber"]])
top_300_visit_genes <- tail(varpart_dvic[["fitted_df"]][top_300_visit_genes_idx, ], n = 300)
summary(top_300_visit_genes)
visit_genes_gp <- simple_gprofiler(rownames(top_300_visit_genes))
visit_genes_gp
```

### Drug

I do not have any idea what one should expect vis a vis the two drugs.
Looking at the following result, I guess one should not be surprised
given that I am unaware of m/any experiments which explicitly compare
antimonial treatments with respect to transcriptional profile.

```{r}
top_300_drug_genes_idx <- order(varpart_dvic[["fitted_df"]][["drug"]])
top_300_drug_genes <- tail(varpart_dvic[["fitted_df"]][top_300_drug_genes_idx, ], n = 300)
summary(top_300_drug_genes)
drug_genes_gp <- simple_gprofiler(rownames(top_300_drug_genes))
drug_genes_gp
drug_genes_gp$pvalue_plots$MF
```

### Infecting species

The violin plot above suggests there is very limited variance with
respect to the two species.  In addition, there are significantly
fewer braziliensis samples which failed treatment (for miltefosine in
particular), so I presume variancePartition is going to have trouble
extracting genes which are reliable in this context.  In addition, I
would assume that the human transcriptome has pretty similar responses
to the two parasites, exacerbating this confounder.

```{r}
top_300_species_genes_idx <- order(varpart_dvic[["fitted_df"]][["infectingspecies"]])
top_300_species_genes <- tail(varpart_dvic[["fitted_df"]][top_300_species_genes_idx, ], n = 300)
summary(top_300_species_genes)
species_genes_gp <- simple_gprofiler(rownames(top_300_species_genes))
species_genes_gp
```

### Clinical outcome

In my mind, clinical outcome is the most generally interesting query.
It is also one with limited information in the data, _but_ lots of
other studies have been performed in this realm, so it is more likely
to have reliable overrepresentation results even with limited
information.  Thus, when I look at the bar/dot plot the categories look
interesting.

The following block provides an example of one of the fun things my
gProfiler2 function does: it coerces the gprofiler output to the
datastructure used by clusterProfiler and thus may be plotted with the
various functions in the 'enrichplot' package.  It also provides a
copy of the cute interactive plots produced by gprofiler.

```{r}
top_300_cf_genes_idx <- order(varpart_dvic[["fitted_df"]][["clinicaloutcome"]])
top_300_cf_genes <- tail(varpart_dvic[["fitted_df"]][top_300_cf_genes_idx, ], n = 300)
summary(top_300_cf_genes)
cf_genes_gp <- simple_gprofiler(rownames(top_300_cf_genes))
cf_genes_gp
enrichplot::dotplot(cf_genes_gp$GO_enrich)
```

# The samples' distribution is similar without 2 samples

Here are the PCA results before/after removing the two samples that I
called shenanigans on.  They are quite similar.  The most notable
difference is the weirdo failure visit 3 sample on the left goes away.

```{r}
wellcome_outtime_norm <- normalize_expt(wellcome_outtime, filter = TRUE,
                                        convert = "cpm", transform = "log2", norm = "quant")
pre_filter_pca <- plot_pca(wellcome_outtime_norm)
pp(file = "images/all_samples_norm_pca.svg")
pre_filter_pca$plot
dev.off()
pre_filter_pca

wellcome_outfilt_norm <- normalize_expt(wellcome_filtered, filter = TRUE,
                                        convert = "cpm", transform = "log2", norm = "quant")
post_filter_pca <- plot_pca(wellcome_outfilt_norm)
pp(file = "images/filtered_samples_norm_pca.svg")
post_filter_pca$plot
dev.off()
post_filter_pca
```

# Visualize Various PCA

Now that we have an initial sense of how the samples relate to each
other, let us poke a bit further.  In each of the following, I am
likely to do one plot before and one after using sva.

## By visit

I think the pattern of samples by visit is pretty or satisfying in a
'hey, these things have some internal sense' way.  I guess modifying
the counts with surrogates from sva makes it a little bit clearer?  It
does increase the PC1 variance a bit I suppose.

```{r}
wellcome_time_norm <- normalize_expt(wellcome_time_filt, norm = "quant", convert = "cpm",
                                     filter = TRUE, transform = "log2")
time_norm_pca <- plot_pca(wellcome_time_norm)
pp(file = "images/time_norm_pca.svg")
time_norm_pca$plot
dev.off()
time_norm_pca

wellcome_time_nb <- normalize_expt(wellcome_time_filt, convert = "cpm",
                                   batch = "svaseq", filter = TRUE, transform = "log2")
time_nb_pca <- plot_pca(wellcome_time_nb)
pp(file = "images/time_nb_pca.svg")
time_nb_pca$plot
dev.off()
time_nb_pca
```

## Cure/Fail

In contrast, the ability of sva to determine variance which is related
to clinical outcome is _very_ limited.  The resulting PCA plot is
a bit of a mess.

```{r}
wellcome_outcome_norm <- normalize_expt(wellcome_outcome_filt, norm = "quant", convert = "cpm",
                                        filter = TRUE, transform = "log2")
outcome_norm_pca <- plot_pca(wellcome_outcome_norm, plot_labels = FALSE)
pp(file = "images/outcome_norm_pca.svg")
outcome_norm_pca$plot
dev.off()
outcome_norm_pca

wellcome_outcome_nb <- normalize_expt(wellcome_outcome_filt, convert = "cpm",
                                      batch = "svaseq", filter = TRUE, transform = "log2")
outcome_nb_pca <- plot_pca(wellcome_outcome_nb)
pp(file = "images/outcome_nb_pca.svg")
outcome_nb_pca$plot
dev.off()
outcome_nb_pca
```

## Drug

At first glance, it at least appears possible that we can get a handle
on differences between the two drug treatment regimes; but I kind of
think that is a trick played by our eyes.  Glancing at the plot from
sva, it appears to disagree with me; at least this recapitulates the
variancePartition result I think.

```{r}
wellcome_drug_norm <- normalize_expt(wellcome_drug_filt, norm = "quant", convert = "cpm",
                                     filter = TRUE, transform = "log2")
drug_norm_pca <- plot_pca(wellcome_drug_norm)
pp(file = "images/drug_norm_pca.svg")
drug_norm_pca$plot
dev.off()
drug_norm_pca

wellcome_drug_nb <- normalize_expt(wellcome_drug_filt, convert = "cpm",
                                   batch = "svaseq", filter = TRUE, transform = "log2")
drug_nb_pca <- plot_pca(wellcome_drug_nb)
pp(file = "images/drug_nb_pca.svg")
drug_nb_pca$plot
dev.off()
drug_nb_pca
```

### Separate drug treatment by time

Conversely, it is likely that we want to separate and/or combine the
various factors with the visit.  In the following block we will use
the factor which is a combination of drug and visit.

```{r}
drugtime <- set_expt_conditions(wellcome_filtered, fact = "drug_visit",
                                     colors = color_choices[["drug_visit"]]) %>%
  set_expt_batches(fact = "clinicaloutcome")

drugtime_norm <- normalize_expt(drugtime, norm = "quant", convert = "cpm",
                                transform = "log2", filter = TRUE)
drugtime_norm_pca <- plot_pca(drugtime_norm)
pp(file = "images/drugtime_norm_pca.svg")
drugtime_norm_pca$plot
dev.off()
drugtime_norm_pca

drugtime_nb <- normalize_expt(drugtime, batch = "svaseq", convert = "cpm",
                              transform = "log2", filter = TRUE)
drugtime_nb_pca <- plot_pca(drugtime_nb)
pp(file = "images/drugtime_nb_pca.svg")
drugtime_nb_pca$plot
dev.off()
drugtime_nb_pca
```

#### Visit 1

One interpretation of some queries from reviewers would be to replot
the various PCAs with the visits separated from each other.  I am not
completely sure that is the correct interpretation, but here it is.

```{r}
drug_v1 <- subset_expt(drugtime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "clinicaloutcome")

drug_v1_norm <- normalize_expt(drug_v1, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
drug_v1_norm_pca <- plot_pca(drug_v1_norm)
pp(file = "images/drug_v1_norm_pca.svg")
drug_v1_norm_pca$plot
dev.off()
drug_v1_norm_pca

drug_v1_nb <- normalize_expt(drug_v1, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
drug_v1_nb_pca <- plot_pca(drug_v1_nb)
pp(file = "images/drug_v1_nb_pca.svg")
drug_v1_nb_pca$plot
dev.off()
drug_v1_nb_pca
```

#### Visit 2

I would like to have something useful to say before every block.  I don't.

```{r}
drug_v2 <- subset_expt(drugtime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "clinicaloutcome")

drug_v2_norm <- normalize_expt(drug_v2, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
drug_v2_norm_pca <- plot_pca(drug_v2_norm)
pp(file = "images/drug_v2_norm_pca.svg")
drug_v2_norm_pca$plot
dev.off()
drug_v2_norm_pca

drug_v2_nb <- normalize_expt(drug_v2, batch = "svaseq", convert = "cpm",
                             filter = TRUE, transform = "log2")
drug_v2_nb_pca <- plot_pca(drug_v2_nb)
pp(file = "images/drug_v2_nb_pca.svg")
drug_v2_nb_pca$plot
dev.off()
drug_v2_nb_pca
```

#### Visit 3

```{r}
drug_v3 <- subset_expt(drugtime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "clinicaloutcome")

drug_v3_norm <- normalize_expt(drug_v3, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
drug_v3_norm_pca <- plot_pca(drug_v3_norm)
pp(file = "images/drug_v3_norm_pca.svg")
drug_v3_norm_pca$plot
dev.off()

drug_v3_nb <- normalize_expt(drug_v3, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
drug_v3_nb_pca <- plot_pca(drug_v3_nb)
pp(file = "images/drug_v3_nb_pca.svg")
drug_v3_nb_pca$plot
dev.off()
drug_v3_nb_pca
```

### Outcome alone

```{r}
outcome <- set_expt_conditions(wellcome_filtered, fact = "clinicaloutcome",
                               colors = color_choices[["clinicaloutcome"]]) %>%
  set_expt_batches(fact = "drug")

outcome_norm <- normalize_expt(outcome, norm = "quant", convert = "cpm",
                               transform = "log2", filter = TRUE)
outcome_norm_pca <- plot_pca(outcome_norm)
pp(file = "images/outcome_norm_pca.svg")
outcome_norm_pca$plot
dev.off()
outcome_norm_pca

outcome_nb <- normalize_expt(outcome, batch = "svaseq", convert = "cpm",
                             transform = "log2", filter = TRUE)
outcome_nb_pca <- plot_pca(outcome_nb)
pp(file = "images/outcome_nb_pca.svg")
outcome_nb_pca$plot
dev.off()
outcome_nb_pca
```

#### Visit 1

```{r}
outcome_v1 <- subset_expt(outcome, subset = "visitnumber=='v1'")

outcome_v1_norm <- normalize_expt(outcome_v1, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
outcome_v1_norm_pca <- plot_pca(outcome_v1_norm)
pp(file = "images/outcome_v1_norm_pca.svg")
outcome_v1_norm_pca$plot
dev.off()
outcome_v1_norm_pca

outcome_v1_nb <- normalize_expt(outcome_v1, batch = "svaseq", convert = "cpm",
                                filter = TRUE, transform = "log2")
outcome_v1_nb_pca <- plot_pca(outcome_v1_nb)
pp(file = "images/outcome_v1_nb_pca.svg")
outcome_v1_nb_pca$plot
dev.off()
outcome_v1_nb_pca
```

#### Visit 2

I would like to have something useful to say before every block.  I don't.

```{r}
outcome_v2 <- subset_expt(outcome, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "clinicaloutcome")

outcome_v2_norm <- normalize_expt(outcome_v2, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
outcome_v2_norm_pca <- plot_pca(outcome_v2_norm)
pp(file = "images/outcome_v2_norm_pca.svg")
outcome_v2_norm_pca$plot
dev.off()
outcome_v2_norm_pca

outcome_v2_nb <- normalize_expt(outcome_v2, batch = "svaseq", convert = "cpm",
                             filter = TRUE, transform = "log2")
outcome_v2_nb_pca <- plot_pca(outcome_v2_nb)
pp(file = "images/outcome_v2_nb_pca.svg")
outcome_v2_nb_pca$plot
dev.off()
outcome_v2_nb_pca
```

#### Visit 3

```{r}
outcome_v3 <- subset_expt(outcome, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "clinicaloutcome")

outcome_v3_norm <- normalize_expt(outcome_v3, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
outcome_v3_norm_pca <- plot_pca(outcome_v3_norm)
pp(file = "images/outcome_v3_norm_pca.svg")
outcome_v3_norm_pca$plot
dev.off()
outcome_v3_norm_pca

outcome_v3_nb <- normalize_expt(outcome_v3, batch = "svaseq", convert = "cpm",
                                filter = TRUE, transform = "log2")
outcome_v3_nb_pca <- plot_pca(outcome_v3_nb)
pp(file = "images/outcome_v3_nb_pca.svg")
outcome_v3_nb_pca$plot
dev.off()
outcome_v3_nb_pca
```

### Outcome and time

Repeat the above process, but this time using clinical outcome and
time combined.

```{r}
wellcome_outtime <- set_expt_conditions(wellcome_filtered, fact = "outcome_visit")

wellcome_outtime_norm <- normalize_expt(wellcome_outtime, norm = "quant", convert = "cpm",
                                        filter = TRUE, transform = "log2")
outtime_norm_pca <- plot_pca(wellcome_outtime_norm)
pp(file = "images/outtime_norm_pca.svg")
outtime_norm_pca$plot
dev.off()
outtime_norm_pca

wellcome_outtime_nb <- normalize_expt(wellcome_outtime, convert = "cpm",
                                      batch = "svaseq", filter = TRUE, transform = "log2")
outtime_nb_pca <- plot_pca(wellcome_outtime_nb)
pp(file = "images/outtime_nb_pca.svg")
outtime_nb_pca$plot
dev.off()
outtime_nb_pca
```

#### Visit 1

```{r}
outtime_v1 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "drug")

outtime_v1_norm <- normalize_expt(outtime_v1, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
outtime_v1_norm_pca <- plot_pca(outtime_v1_norm)
pp(file = "images/outtime_v1_norm_pca.svg")
outtime_v1_norm_pca$plot
dev.off()
outtime_v1_norm_pca$plot

outtime_v1_nb <- normalize_expt(outtime_v1, batch = "svaseq", convert = "cpm",
                                filter = TRUE, transform = "log2")
outtime_v1_nb_pca <- plot_pca(outtime_v1_nb)
pp(file = "images/outtime_v1_nb_pca.svg")
outtime_v1_nb_pca
```

#### Visit 2

```{r}
outtime_v2 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "drug")

outtime_v2_norm <- normalize_expt(outtime_v2, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
outtime_v2_norm_pca <- plot_pca(outtime_v2_norm)
pp(file = "images/outtime_v2_norm_pca.svg")
outtime_v2_norm_pca$plot
dev.off()
outtime_v2_norm_pca$plot

outtime_v2_nb <- normalize_expt(outtime_v2, batch = "svaseq", convert = "cpm",
                                filter = TRUE, transform = "log2")
outtime_v2_nb_pca <- plot_pca(outtime_v2_nb)
pp(file = "images/outtime_v2_nb_pca.svg")
outtime_v2_nb_pca
```

#### Visit 3

```{r}
outtime_v3 <- subset_expt(wellcome_outtime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "drug")

outtime_v3_norm <- normalize_expt(outtime_v3, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
outtime_v3_norm_pca <- plot_pca(outtime_v3_norm)
pp(file = "images/outtime_v3_norm_pca.svg")
outtime_v3_norm_pca$plot
dev.off()
outtime_v3_norm_pca$plot

outtime_v3_nb <- normalize_expt(outtime_v3, batch = "svaseq", convert = "cpm",
                                filter = TRUE, transform = "log2")
outtime_v3_nb_pca <- plot_pca(outtime_v3_nb)
pp(file = "images/outtime_v3_nb_pca.svg")
outtime_v3_nb_pca
```

## Parasite Species

### Separate species by time

```{r}
speciestime <- set_expt_conditions(wellcome_outtime, fact = "species_visit",
                                   colors = color_choices[["species_visit"]]) %>%
  set_expt_batches(fact = "clinicaloutcome")

speciestime_norm <- normalize_expt(speciestime, norm = "quant", convert = "cpm",
                                   transform = "log2", filter = TRUE)
speciestime_norm_pca <- plot_pca(speciestime_norm)
pp(file = "images/speciestime_norm_pca.svg")
speciestime_norm_pca$plot
dev.off()
speciestime_norm_pca

speciestime_nb <- normalize_expt(speciestime, batch = "svaseq", convert = "cpm",
                                 transform = "log2", filter = TRUE)
speciestime_nb_pca <- plot_pca(speciestime_nb)
pp(file = "images/speciestime_nb_pca.svg")
speciestime_nb_pca$plot
dev.off()
speciestime_nb_pca
```

#### Visit 1

```{r}
species_v1 <- subset_expt(speciestime, subset = "visitnumber=='v1'") %>%
  set_expt_batches(fact = "clinicaloutcome")

species_v1_norm <- normalize_expt(species_v1, norm = "quant", convert = "cpm",
                               filter = TRUE, transform = "log2")
species_v1_norm_pca <- plot_pca(species_v1_norm)
pp(file = "images/species_v1_norm_pca.svg")
species_v1_norm_pca$plot
dev.off()
species_v1_norm_pca

species_v1_nb <- normalize_expt(species_v1, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
species_v1_nb_pca <- plot_pca(species_v1_nb)
pp(file = "images/species_v1_nb_pca.svg")
species_v1_nb_pca$plot
dev.off()
species_v1_nb_pca
```

#### Visit 2

```{r}
species_v2 <- subset_expt(speciestime, subset = "visitnumber=='v2'") %>%
  set_expt_batches(fact = "clinicaloutcome")

species_v2_norm <- normalize_expt(species_v2, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
species_v2_norm_pca <- plot_pca(species_v2_norm)
pp(file = "images/species_v2_norm_pca.svg")
species_v2_norm_pca$plot
dev.off()
species_v2_norm_pca

species_v2_nb <- normalize_expt(species_v2, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
species_v2_nb_pca <- plot_pca(species_v2_nb)
pp(file = "images/species_v2_nb_pca.svg")
species_v2_nb_pca$plot
dev.off()
species_v2_nb_pca
```

#### Visit 3

```{r}
species_v3 <- subset_expt(speciestime, subset = "visitnumber=='v3'") %>%
  set_expt_batches(fact = "clinicaloutcome")

species_v3_norm <- normalize_expt(species_v3, norm = "quant", convert = "cpm",
                                  filter = TRUE, transform = "log2")
species_v3_norm_pca <- plot_pca(species_v3_norm)
pp(file = "images/species_v3_norm_pca.svg")
species_v3_norm_pca$plot
dev.off()

species_v3_nb <- normalize_expt(species_v3, batch = "svaseq", convert = "cpm",
                               filter = TRUE, transform = "log2")
species_v3_nb_pca <- plot_pca(species_v3_nb)
pp(file = "images/species_v3_nb_pca.svg")
species_v3_nb_pca$plot
dev.off()
species_v3_nb_pca
```

#### Back to everything

```{r}
wellcome_parasite_norm <- normalize_expt(wellcome_parasite_filt, norm = "quant", convert = "cpm",
                                         filter = TRUE, transform = "log2")
plot_pca(wellcome_parasite_norm)

wellcome_parasite_nb <- normalize_expt(wellcome_parasite_filt, norm = "quant", convert = "cpm",
                                       batch = "svaseq", filter = TRUE, transform = "log2")
plot_pca(wellcome_parasite_nb)
```

Ok, so that is a big pile of pictures, now let us perform some DE
analyses and see what pops out.

# DE analyses

Given the above variance partition and PCA plots, we can surmise that
some metadata factors are much more likely to give interesting results
than others (Drug far more than C/F, for example).  With that in mind,
let us perform the various possible DE analyses with each
factor and see what comes out.  Each of these runs of all pairwise
will be performed once with SVA estimates in the model, and once with
the drug treatment as the batch factor.

## Outcome and time

One thing to recall, my sanitizer removes punctuation from
concatenated factors; so 'cure_v2' turns into 'curev2'.

## The contrasts of interest

The following list names each contrast and the two element vector
following provides each numerator and denominator.

```{r}
outtime_keepers <- list(
  "curev1v2" = c("curev2", "curev1"),
  "curev1v3" = c("curev3", "curev1"),
  "curev2v3" = c("curev3", "curev2"),
  "failv1v2" = c("failurev2", "failurev1"),
  "failv1v3" = c("failurev3", "failurev1"),
  "failv2v3" = c("failurev3", "failurev2"),
  "cfv1" = c("failurev1", "curev1"),
  "cfv2" = c("failurev2", "curev2"),
  "cfv3" = c("failurev3", "curev3"))
```

Start with the combined factor of {outcome}_{visit}.  Given the PCA, I
expect to see a little bit of information with batch in the model,
oddly less information with SVA.

### SVA

One of the review comments seemed to me to suggest that using the
variancePartition dream method might be useful.  I am not certain if
that is currently enabled.  I definitely got it working, though.  Also
note that filter in this context is on a gene basis, not sample.

```{r}
outtime_sva_de <- all_pairwise(wellcome_filtered, model_batch = "svaseq", filter = TRUE)
outtime_sva_de

outtime_sva_table <- combine_de_tables(
  outtime_sva_de,
  keepers = outtime_keepers,
  excel = glue("excel/CL_biopsies_outtime_table_sva-v{ver}.xlsx"))
outtime_sva_table

outtime_sva_sig <- extract_significant_genes(
  outtime_sva_table,
  excel = glue("excel/CL_biopsies_outtime_sig_sva-v{ver}.xlsx"))
outtime_sva_sig
```

### Batch in model

```{r}
outtime_batch_de <- all_pairwise(wellcome_filtered, model_batch = TRUE, filter = TRUE)
outtime_batch_de

outtime_batch_table <- combine_de_tables(
  outtime_batch_de,
  keepers = outtime_keepers,
  excel = glue("excel/CL_biopsies_outtime_table_batch-v{ver}.xlsx"))
outtime_batch_table

outtime_batch_sig <- extract_significant_genes(
  outtime_batch_table,
  excel = glue("excel/CL_biopsies_outtime_sig_batch-v{ver}.xlsx"))
outtime_batch_sig
```

# PROPER

There was a request some time ago to provide a power analysis.  I
arbitrarily chose this dataset in order to invoke PROPER and provide
the resulting plots...

```{r}
outtime_batch_proper <- simple_proper(outtime_batch_table)
outtime_batch_proper[["curev2_vs_curev1"]][["power_table"]]
outtime_batch_proper[["curev2_vs_curev1"]][["power_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powertd_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powerfd_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["fdcost_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["powerhist_plot"]]
outtime_batch_proper[["curev2_vs_curev1"]][["poweralpha_plot"]]
```

## Only time

Now let us compare the time points, with and without SVA.

## Time contrasts

```{r}
time_keepers <- list(
  "v1v2" = c("v2", "v1"),
  "v1v3" = c("v3", "v1"),
  "v2v3" = c("v3", "v2"))
```

### SVA

```{r}
time_sva_de <- all_pairwise(wellcome_time_filt, model_batch = "svaseq", filter = TRUE)
time_sva_de

time_sva_table <- combine_de_tables(
  time_sva_de,
  keepers = time_keepers,
  excel = glue("excel/CL_biopsies_time_table_sva-v{ver}.xlsx"))
time_sva_table

time_sva_sig <- extract_significant_genes(
  time_sva_table,
  excel = glue("excel/CL_biopsies_time_sig_sva-v{ver}.xlsx"))
time_sva_sig
```

### Batch

```{r}
time_batch_de <- all_pairwise(wellcome_time_filt, model_batch = "batchseq", filter = TRUE)
time_batch_de

time_batch_table <- combine_de_tables(
  time_batch_de,
  keepers = time_keepers,
  excel = glue("excel/CL_biopsies_time_table_batch-v{ver}.xlsx"))
time_batch_table

time_batch_sig <- extract_significant_genes(
  time_batch_table,
  excel = glue("excel/CL_biopsies_time_batch_sva-v{ver}.xlsx"))
time_batch_sig
```

## Strain

I neglected to set a contrast name/value for this.

```{r}
## Oh, I just let the computer choose the numerator/denominator on its own.
```

The infecting strain I think should prove one of the more interesting
comparisons.

### SVA

```{r}
parasite_sva_de <- all_pairwise(wellcome_parasite_filt, model_batch = "svaseq", filter = TRUE)
parasite_sva_de

parasite_sva_table <- combine_de_tables(
  parasite_sva_de,
  excel = glue("excel/CL_biopsies_parasite_table_sva-v{ver}.xlsx"))
parasite_sva_table

parasite_sva_sig <- extract_significant_genes(
  parasite_sva_table,
  excel = glue("excel/CL_biopsies_parasite_sig_sva-v{ver}.xlsx"))
parasite_sva_sig
```

### Batch in model

```{r}
parasite_batch_de <- all_pairwise(wellcome_parasite_filt, model_batch = TRUE, filter = TRUE)
parasite_batch_de

parasite_batch_table <- combine_de_tables(
  parasite_batch_de,
  excel = glue("excel/CL_biopsies_parasite_table_batch-v{ver}.xlsx"))
parasite_batch_table

parasite_batch_sig <- extract_significant_genes(
  parasite_batch_table,
  excel = glue("excel/CL_biopsies_parasite_sig_sva-v{ver}.xlsx"))
parasite_batch_sig
```

## Outcome

As stated above, clinical outcome by itself does not have much
information content in this dataset.

### SVA

```{r}
outcome_sva_de <- all_pairwise(wellcome_outcome_filt, model_batch = "svaseq", filter = TRUE)
outcome_sva_de

outcome_sva_table <- combine_de_tables(
  outcome_sva_de,
  excel = glue("excel/CL_biopsies_outcome_table_sva-v{ver}.xlsx"))
outcome_sva_table

outcome_sva_sig <- extract_significant_genes(
  outcome_sva_table,
  excel = glue("excel/CL_biopsies_outcome_sig_sva-v{ver}.xlsx"))
outcome_sva_sig
```

### Batch in model

```{r}
outcome_batch_de <- all_pairwise(wellcome_outcome_filt, model_batch = TRUE, filter = TRUE)
outcome_batch_de

outcome_batch_table <- combine_de_tables(
  outcome_batch_de,
  excel = glue("excel/CL_biopsies_outcome_table_batch-v{ver}.xlsx"))
outcome_batch_table

outcome_batch_sig <- extract_significant_genes(
  outcome_batch_table,
  excel = glue("excel/CL_biopsies_outcome_sig_sva-v{ver}.xlsx"))
outcome_batch_sig
```

## Drug treatment

### SVA

```{r}
drug_sva_de <- all_pairwise(wellcome_drug_filt, model_batch = "svaseq", filter = TRUE)
drug_sva_de

drug_sva_table <- combine_de_tables(
  drug_sva_de,
  excel = glue("excel/CL_biopsies_drug_table_sva-v{ver}.xlsx"))
drug_sva_table

drug_sva_sig <- extract_significant_genes(
  drug_sva_table,
  excel = glue("excel/CL_biopsies_drug_sig_sva-v{ver}.xlsx"))
drug_sva_sig
```

### Batch in model

```{r}
drug_batch_de <- all_pairwise(wellcome_drug_filt, model_batch = TRUE, filter = TRUE)
drug_batch_de

drug_batch_table <- combine_de_tables(
  drug_batch_de,
  excel = glue("excel/CL_biopsies_drug_table_batch-v{ver}.xlsx"))
drug_batch_table

drug_batch_sig <- extract_significant_genes(
  drug_batch_table,
  excel = glue("excel/CL_biopsies_drug_sig_sva-v{ver}.xlsx"))
drug_batch_sig
```

## Parasite and time

This requires setting up the contrasts of interest.  I am going to
assume cross species/cross time are not interesting, and that v3/v1
and v3/v2 are not interesting.

Note to self, do not forget that I sanitize the data by removing punctuation!

```{r}
speciestime_keepers <- list(
  "pan_v2v1" = c("lvpanamensisv2", "lvpanamensisv1"),
  "pan_v3v1" = c("lvpanamensisv3", "lvpanamensisv1"),
  "braz_v2v1" = c("lvbraziliensisv2", "lvbraziliensisv1"),
  "braz_v3v1" = c("lvbraziliensisv3", "lvbraziliensisv1"),
  "v1_panbraz" = c("lvbraziliensisv1", "lvpanamensisv1"),
  "v2_panbraz" = c("lvbraziliensisv2", "lvpanamensisv2"),
  "v3_panbraz" = c("lvbraziliensisv3", "lvpanamensisv3"))
```

### SVA

```{r}
speciestime_sva_de <- all_pairwise(speciestime, model_batch = "svaseq", filter = TRUE)
speciestime_sva_de

speciestime_sva_table <- combine_de_tables(
  speciestime_sva_de, keepers = speciestime_keepers,
  excel = glue("excel/CL_biopsies_speciestime_table_sva-v{ver}.xlsx"))
speciestime_sva_table

speciestime_sva_sig <- extract_significant_genes(
  speciestime_sva_table,
  excel = glue("excel/CL_biopsies_speciestime_sig_sva-v{ver}.xlsx"))
speciestime_sva_sig
```

### Batch in model

```{r}
speciestime_sva_de <- all_pairwise(speciestime, model_batch = TRUE, filter = TRUE)
speciestime_sva_de

speciestime_sva_table <- combine_de_tables(
  speciestime_sva_de, keepers = speciestime_keepers,
  excel = glue("excel/CL_biopsies_speciestime_table_sva-v{ver}.xlsx"))
speciestime_sva_table

speciestime_sva_sig <- extract_significant_genes(
  speciestime_sva_table,
  excel = glue("excel/CL_biopsies_speciestime_sig_sva-v{ver}.xlsx"))
speciestime_sva_sig
```

## Drug treatment and time

Same rules here as strain+time

```{r}
drugtime_keepers <- list(
  "milt_v2v1" = c("miltefosinev2", "miltefosinev1"),
  "milt_v3v1" = c("miltefosinev3", "miltefosinev1"),
  "gluc_v2v1" = c("antimoniatev2", "antimoniatev1"),
  "gluc_v3v1" = c("antimoniatev3", "antimoniatev1"),
  "v1_miltgluc" = c("miltefosinev1", "antimoniatev1"),
  "v2_miltgluc" = c("miltefosinev2", "antimoniatev2"),
  "v3_miltgluc" = c("miltefosinev3", "antimoniatev3"))
```

### SVA

```{r}
drugtime_sva_de <- all_pairwise(drugtime, model_batch = "svaseq", filter = TRUE)
drugtime_sva_de

drugtime_sva_table <- combine_de_tables(
  drugtime_sva_de, keepers = drugtime_keepers,
  excel = glue("excel/CL_biopsies_drugtime_table_sva-v{ver}.xlsx"))
drugtime_sva_table

drugtime_sva_sig <- extract_significant_genes(
  drugtime_sva_table,
  excel = glue("excel/CL_biopsies_drugtime_sig_sva-v{ver}.xlsx"))
drugtime_sva_sig
```

### Batch in model

```{r}
drugtime_sva_de <- all_pairwise(drugtime, model_batch = TRUE, filter = TRUE)
drugtime_sva_de

drugtime_sva_table <- combine_de_tables(
  drugtime_sva_de, keepers = drugtime_keepers,
  excel = glue("excel/CL_biopsies_drugtime_table_sva-v{ver}.xlsx"))
drugtime_sva_table

drugtime_sva_sig <- extract_significant_genes(
  drugtime_sva_table,
  excel = glue("excel/CL_biopsies_drugtime_sig_sva-v{ver}.xlsx"))
drugtime_sva_sig
```

# Ontology searches

Given the results in the previous blocks, let us use gProfiler2 in
look for over represented categories/ontologies in the suspect
databases: GO/reactome/kegg/TF/wikipath/etc.

I think for the moment I will do this entirely with the sva results.

## Outcome and time

I have a series of wrapper functions for
gProfiler2/goseq/clusterProfiler/topGO/GOstats.  Some of them are
smart enough to take as input the result from
extract_significant_genes().

There is also a function which writes pretty-ified ontology results,
but it works on a single-table basis.

```{r}
outtime_gprofiler <- all_gprofiler(outtime_sva_sig)

outtime_gprofiler$curev1v2_up$pvalue_plots$BP
outtime_gprofiler$failv1v2_up$pvalue_plots$KEGG
outtime_gprofiler$cfv1_up$pvalue_plots$REAC

## Note, I am writing a set of methods for write_gprofiler_data so that it is
## aware of the result from all_gprofiler and can handle either all contrasts
## or a set of desired contrasts.
write_all <- function(result) {
  for (n in names(result)) {
    name <- glue("excel/gprofiler_{n}.xlsx")
    res <- result[[n]]
    written <- try(write_gprofiler_data(res, excel = name))
  }
}

written <- write_all(outtime_gprofiler)
```

## Visit

```{r}
visit_gprofiler <- all_gprofiler(time_sva_sig)

written <- write_all(visit_gprofiler)
```

## Strain

```{r}
parasite_gprofiler <- all_gprofiler(parasite_sva_sig)

written <- write_all(parasite_gprofiler)
```

## Outcome

```{r}
outcome_gprofiler <- all_gprofiler(outcome_sva_sig)

written <- write_all(outcome_gprofiler)
```

## Treatment

```{r}
drug_gprofiler <- all_gprofiler(drug_sva_sig)

written <- write_all(drug_gprofiler)
```

## Strain and time

```{r}
speciestime_gprofiler <- all_gprofiler(speciestime_sva_sig)

written <- write_all(speciestime_gprofiler)
```

## Drug and time

```{r}
drugtime_gprofiler <- all_gprofiler(drugtime_sva_sig)

written <- write_all(drugtime_gprofiler)
```

# GSVA

From my perspective, it seems that the most interesting GSVA
comparison is to look for scores changing between the cure and fail
samples.  The following block therefore passes the raw expression data
and performs the gene set variance analysis of it against the mSigDB
C2 set of categories by default.  We can pretty easily change the
mSigDB release/category set; except I would need download the dataset
to the container to use a newer revision and I am not sure about
potential licensing restrictions; so this will just use the publicly
available signatures in the 'GSVAdata' package.

```{r}
wt_gsva <- simple_gsva(wellcome_outcome_filt, signature_category = "c7")
wt_gsva

wt_gsva_sig <- get_sig_gsva_categories(wt_gsva, excel = "excel/CL_biopsies_gsva_outcome_c7.xlsx")
wt_gsva_sig
```

Hmm, did I get confused, is C7 what I thought it was?  Yeah, it is the
immunologic signature sets.  C2 is the larger set of experiments.

```{r}
wellcome_gsva_c2 <- simple_gsva(wellcome_filtered, signature_category = "c2")
wellcome_gsva_c2_sig <- get_sig_gsva_categories(
  wellcome_gsva_c2,
  excel = "excel/CL_biopsies_gsva_outcome_c2.xlsx")
```

The following blocks may be used to save and reload the state of the data.

```{r}
pander::pander(sessionInfo())
message("This is hpgltools commit: ", get_git_commit())
this_save <- paste0(gsub(pattern = "\\.Rmd", replace = "", x = rmd_file), "-v", ver, ".rda.xz")
#message("Saving to ", this_save)
tmp <- sm(saveme(filename = this_save))
```

```{r loadme, eval=FALSE}
loadme(filename = this_save)
```
